### PR TITLE
Add a sample that uses the iree_codegen.ukernel.generic operation to call a microkernel supplied as a plugin.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
This PR adds a sample that mirrors the `standalone_plugin` and
`system_plugin` examples, but instaed of using a stream-level shims
for invoking the micro kernel, uses the `iree_codegen.ukernel.generic`
to call a specific microkernel. The input IR contains a dispatch
(expressed within `flow.dispatch.region`) which extracts slices of the
inputs/outputs on a per-workgroup basis and calls the microkernels
provided.

Depends on https://reviews.llvm.org/D148684

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
@benvanik based on discussion last week started creating a sample [here](https://github.com/openxla/iree/pull/13132/files#diff-f366ebe2f1ce6ae6afce3b483a5d6e35f5fc4ccea608f416a72688abe76b456f) that uses the `iree_codegen.ukernel.generic` operation. It compiles as I expect (see gist [here](https://gist.githubusercontent.com/MaheshRavishankar/2c497f71bb5dc2f9531120a7ccb7b80e/raw/d3cb7fcd9949137d99a94a9f4138d70beefac337/tmp_ukernel.dump.mlir)) . Now I am trying to see what other wrappers I need to add. Question for you is which example in https://github.com/openxla/iree/tree/main/samples/custom_dispatch/cpu/plugin should I follow?

**Comment by benvanik** [comment]
Wow, this looks awesome!

Both samples are identical IR split only to allow for optional enabling of either loading mechanism so we should be able to make both of them use what you have here. If you want to land what's working I'm happy to rework them, but the hope is to replace the flow.dispatch https://github.com/openxla/iree/blob/main/samples/custom_dispatch/cpu/plugin/standalone_example.mlir#L105 with your flow.dispatch.workgroups + ukernel op and delete the stream.executable entirely.

**Comment by MaheshRavishankar** [comment]
> Wow, this looks awesome!
> 
> Both samples are identical IR split only to allow for optional enabling of either loading mechanism so we should be able to make both of them use what you have here. If you want to land what's working I'm happy to rework them, but the hope is to replace the flow.dispatch [`main`/samples/custom_dispatch/cpu/plugin/standalone_example.mlir#L105](https://github.com/openxla/iree/blob/main/samples/custom_dispatch/cpu/plugin/standalone_example.mlir?rgh-link-date=2023-04-17T23%3A33%3A38Z#L105) with your flow.dispatch.workgroups + ukernel op and delete the stream.executable entirely.

Ok, let me just try to do that then.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir [hunk 1]
```diff
@@ -411,9 +411,72 @@ hal.executable private @mmt4d_ukernel {
...
   }
 }
-//      CHECK: func.func private @[[UKERNELFUNC:.+]](
-// CHECK-SAME:     hal.import.static = true
-// CHECK-SAME:     llvm.bareptr
-// CHECK-SAME:     vm.import.module = "vmvx"
-//      CHECK: func @ukernel_dispatch()
-//      CHECK:   func.call @[[UKERNELFUNC]]
+// CHECK-LABEL: func @ukernel_dispatch()
+//       CHECK:   iree_codegen.ukernel.mmt4d
```
Comment (bjacob) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir]: Needs rebasing, I just merged the PR that drops `ukernel.mmt4d`.

File: samples/custom_dispatch/cpu/plugin/standalone_ukernel.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Demonstrates a standalone plugin calling a `simple_mul_workgroup`
+// micro-kernel function. `standalone_ukernel.mlir` shows the input
+// IR to `iree-compile`. The function `ukernel_example` there is
+// compiled to `llvm-cpu` which in turn calls the `simple_mul_workgroup`
+// function defined here. This models calling micro kernels from within
+// dispatches. This standalone plugin mechanism is used here (and barrowed from
+// standalone_plugin.c).
```
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/standalone_ukernel.c]: typo
```suggestion
// dispatches. This standalone plugin mechanism is used here (and borrowed from
// standalone_plugin.c).
```

File: third_party/llvm-project
Comment (ScottTodd) [inline third_party/llvm-project:1]: Reminder to revert changes to `third_party/llvm-project` / wait for the upstream change(s) to land before merging this PR.

File: samples/custom_dispatch/cpu/plugin/standalone_ukernel.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Demonstrates a standalone plugin calling a `simple_mul_workgroup`
+// micro-kernel function. `standalone_ukernel.mlir` shows the input
```
Comment (ScottTodd) [inline samples/custom_dispatch/cpu/plugin/standalone_ukernel.c]: Could add a section to https://github.com/openxla/iree/blob/main/samples/custom_dispatch/cpu/plugin/README.md explaining the different sample versions now in this directory (features, benefits/tradeoffs)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir]: Yup. Also needs an upstream change. So I need to land that before I can land this. But mostly ready for review.

File: third_party/llvm-project
Comment (MaheshRavishankar) [reply third_party/llvm-project:1]: Thanks. Just sent it out to get reviews. WIll land change upstream, wait for it to propagate before landing this PR.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 4]
```diff
@@ -2012,27 +2012,30 @@ static LogicalResult setTranslationInfoAndRootConfig(
...
-          TargetMLTransformInfo::getTargetMLTransformInfo(targetAttr);
-      if (failed(setRootConfigImpl(entryPointFn, rootOperation,
-                                   targetMLTransInfo))) {
+  // Handle the case with no known root operation.
+  if (!rootOperation) {
+    // If there is no translation info set, just set it to default.
+    if (!getTranslationInfo(entryPointFn)) {
+      auto translationInfo = IREE::Codegen::TranslationInfoAttr::get(
+          entryPointFn->getContext(), DispatchLoweringPassPipeline::CPUDefault);
+      // Fall back, just set the translation to CPUDefault.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2024]: maybe can merge this comment to l.2017. I.e., `If there is ..., just set it to CPUDefault.`

File: compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp [hunk 5]
```diff
@@ -74,8 +74,7 @@ static LogicalResult getTileAndDistributeConfig(
     }
   }
   if (!rootOp) {
-    // Just return. All the in-out vectors are empty that should default
-    // the number of workgroups to {1, 1, 1}
+    // THere is no lowering configuration. Return `null`.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp:79]: typo: `There`

**Review by hanhanW — COMMENTED** [review]
Really awesome, just few nits!

**Review by benvanik — CHANGES_REQUESTED** [review]
to keep the samples simpler we should unify them so they at least share the same plugin - the IR can have examples for both ukernel and func paths, but we shouldn't need different c code for them (you can change the c code to match what you need for the ukernel and change the func.calls to match the convention you need)

File: samples/custom_dispatch/cpu/plugin/standalone_ukernel.c (hunk 3 above)
Comment (MaheshRavishankar) [reply samples/custom_dispatch/cpu/plugin/standalone_ukernel.c]: I dont think I need to update the README now. The readme is just describing how the plugin is to be used. THis is just two different tests.

**Review by benvanik — APPROVED** [review]

**Pull request merged**